crates/objc2/src/__macro_helpers/availability.rs

```diff
@@ -0,0 +1,384 @@
+                let digit = match ascii {
+                    b'0'..=b'9' => ascii - b'0',
+                    _ => break,
+                };
+
+                bytes = rest;
+
+                // This handles leading zeroes as well.
+                match ret.checked_mul(10) {
+                    Some(val) => match val.checked_add(digit as _) {
+                        Some(val) => ret = val,
+                        None => panic!("version is too large"),
+                    },
+                    None => panic!("version is too large"),
+                };
+            }
+
+            (ret, bytes)
+        }
+
+        let bytes = version.as_bytes();
+
+        let (major, bytes) = parse_usize(bytes);
+        if major > u16::MAX as usize {
+            panic!("major version is too large");
+        }
+        let major = major as u16;
+
+        let bytes = if let Some((period, bytes)) = bytes.split_first() {
+            if *period != b'.' {
+                panic!("expected period between major and minor version")
+            }
+            bytes
+        } else {
+            return Self {
+                major,
+                minor: 0,
+                patch: 0,
+            };
+        };
+
+        let (minor, bytes) = parse_usize(bytes);
+        if minor > u8::MAX as usize {
+            panic!("minor version is too large");
+        }
+        let minor = minor as u8;
+
+        let bytes = if let Some((period, bytes)) = bytes.split_first() {
+            if *period != b'.' {
+                panic!("expected period after minor version")
+            }
+            bytes
+        } else {
+            return Self {
+                major,
+                minor,
+                patch: 0,
+            };
+        };
+
+        let (patch, bytes) = parse_usize(bytes);
+        if patch > u8::MAX as usize {
+            panic!("patch version is too large");
+        }
+        let patch = patch as u8;
+
+        if !bytes.is_empty() {
+            panic!("too many parts to version");
+        }
+
+        Self {
+            major,
+            minor,
+            patch,
+        }
+    }
+
+    /// Look up the current version at runtime.
+    ///
+    /// The version is cached for performance.
+    #[inline]
+    #[cfg(target_vendor = "apple")]
+    fn current() -> Self {
+        // This is mostly a re-implementation of:
+        // <https://github.com/llvm/llvm-project/blob/llvmorg-19.1.1/compiler-rt/lib/builtins/os_version_check.c>
+        //
+        // Note that this doesn't work with [zippered] `dylib`s yet, though
+        // that's probably fine, `rustc` doesn't support those either.
+        // [zippered]: https://github.com/nico/hack/blob/be6e1a6885a9d5179558b37a0b4c36bec9c4d377/notes/catalyst.md#building-a-zippered-dylib
+
+        // Rust does not yet support weak externs, and besides, I suspect it'd
+        // be dangerous to use, see:
+        // - https://reviews.llvm.org/D150397
```

Comment:
Seems that Flutter has [found a workaround](https://github.com/flutter/flutter/issues/133203), and one that won't need to exist after another XCode toolchain update? As long as there's a fallback, using a better `dlsym` implementation should be fine and keep the improved perf?

Reply:
I actually ended up looking into the implementation of `_availability_version_check`, and found out that `sysctl kern.osproductversion` (or `sysctl kern.iossupportversion` for Mac Catalyst) exists, so I've gone with trying those first instead, and only falling back to the plist on simulators and older platforms.